File: bfs/new/binary_tree_zigzag_level_order_traversal.java
// Here is mine BFS solution

class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        
        if(root == null) {
            return result;
        }
        
        queue.add(root);
        
        while(!queue.isEmpty()) {
            int nodes = queue.size();
            List<Integer> levelNodes = new ArrayList<>();
            
            for(int i = 0; i < nodes; i++) {
                TreeNode current = queue.poll();

                if(result.size() % 2 == 0) {
                    levelNodes.add(current.val);   
                }
                else {
                    levelNodes.add(0, current.val);
                }
                                
                if(current.left != null) {
                    queue.add(current.left);
                }

                if(current.right != null) {
                    queue.add(current.right);
                }
            }
            result.add(levelNodes);
        }
        return result;
     }
}

File: bfs/new/binary_tree_zigzag_level_order_traversal.py
# We can solve this problem by using BFS with queue. Level information is needed in order to reverse the odd row.

def zigzagLevelOrder(self, root):
    res, queue = [], [(root, 0)]
    while queue:
        curr, level = queue.pop(0)
        if curr:
            if len(res) < level+1:
                res.append([])
            if level % 2 == 0:
                res[level].append(curr.val)
            else:
                res[level].insert(0, curr.val)
            queue.append((curr.left, level+1))
            queue.append((curr.right, level+1))
    return res